enable data type propagation

zhuhaozhe opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98065

Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

zhuhaozhe committed:
enable data type propogation

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[WIP] enable data type propagation"


Attach data type information inside scheduling nodes.
Just leverage CI to check whether this attaching will impact UT now.
Plan to do:
(1) Understand and resolve data type propagation related to "masked".
(2) Add debug print to check the data types after data type propagation.
(3) Write UT based on (1).




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[WIP] enable data type propagation"


Attach data type information inside scheduling nodes.
Just leverage CI to check whether this attaching will impact UT now.
Plan to do:
(1) Understand and resolve data type propagation related to "masked".
(2) Add debug print to check the data types after data type propagation.
(3) Write UT based on (1).




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "[WIP] enable data type propagation"


Attach data type information inside scheduling nodes.
Just leverage CI to check whether this attaching will impact UT now.
Plan to do:
(1) Understand and resolve data type propagation related to "masked".
(2) Add debug print to check the data types after data type propagation.
(3) Write UT based on (2).




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[WIP] enable data type propagation"


Attach data type information inside scheduling nodes.
Just leverage CI to check whether this attaching will impact UT now.
Plan to do:
(1) Understand and resolve data type propagation related to "masked".
(2) Add debug print to check the data types after data type propagation.
(3) Write UT based on (2).




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[WIP] enable data type propagation"


Attach data type information inside scheduling nodes.
Just leverage CI to check whether this attaching will impact UT now.
Plan to do:
(1) Understand and resolve data type propagation related to "masked".
(2) Add debug print to check the data types after data type propagation.
(3) Write UT based on (2).




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented:
Have you validated this PR with E2E models?

zhuhaozhe committed:
Update on "enable data type propagation"


Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe commented:
> Have you validated this PR with E2E models?

I verified on a tiny model and some other ut to observe the dtype attached.

jansel reviewed (changes requested):
I don't think this needs to be specific to cpp codegen.  Let's move it to its own file.

zhuhaozhe committed:
Update on "enable data type propagation"


Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable data type propagation"


Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe commented:
> I don't think this needs to be specific to cpp codegen. Let's move it to its own file.

Thanks, I removed it to `common.py`and invoke it in `cpp` path.

jgong5 commented on torch/_inductor/codegen/common.py:
```diff
@@ -36,6 +44,114 @@ def index_prevent_reordering(index: typing.List[sympy.Expr], index_vars, sizes):
+            )
+        data_type_logger(msg + input_msg)
+        _node.meta[OptimizationContext.key] = opt_ctx
+        opt_ctx.dtype_propogated = True
```
Is this a duplicate of `opt_ctx.dtype`? Can we check just check if `opt_ctx.dtype` is `None` or not?

zhuhaozhe replied:
Yes, I removed this.

jgong5 commented on torch/_inductor/codegen/common.py:
```diff
@@ -36,6 +44,114 @@ def index_prevent_reordering(index: typing.List[sympy.Expr], index_vars, sizes):
+            "bitwise_right_shift",
+        ]
+        ops_with_dtype_arg = ["constant", "to_dtype", "rand", "randn"]
+        reduction_to_bool = ["any"]
+        reduction_to_int64 = ["argmin", "argmax"]
```
Simpler to build a dict from reduction type to dtype?

zhuhaozhe replied:
Done

jgong5 commented on torch/_inductor/codegen/common.py:
```diff
@@ -36,6 +44,114 @@ def index_prevent_reordering(index: typing.List[sympy.Expr], index_vars, sizes):
+            "eq",
+            "ne",
+            "bitwise_not",
+            "bitwise_or",
```
I suppose the result can be either bool or integer? Why it has to be bool?
https://pytorch.org/docs/stable/generated/torch.bitwise_or.html

zhuhaozhe replied:
Thanks for reminding here, I removed `bitwise_not, bitwise_or, bitwise_left_shift, bitwise_right_shift` here and also add UT for this.

jgong5 commented on torch/_inductor/codegen/common.py:
```diff
@@ -36,6 +44,114 @@ def index_prevent_reordering(index: typing.List[sympy.Expr], index_vars, sizes):
+def _data_type_propagation(sub_graph: torch.fx.Graph):
+    def propagate_node(node: torch.fx.Node):
+        _node: torch.fx.Node = node
+        if _node.target == "ops":
+            return False
```
Check target in `ops_without_dtype` instead? Then, you don't have to explicitly check `ops_without_dtype` later?

zhuhaozhe replied:
Yes, we can check `target in ops_without_dtype` here, but we still need to explicitly check ops_without_dtype later while propagate dtype from input nodes. 
```
        input_nodes = [
            n
            for n in inputs
            if isinstance(n, torch.fx.node.Node) and n.target not in ops_without_dtype
        ]
```
While `node.inputs.target = ['ops' ,'get_index', 'add'],` we need remove `ops` and `get_index` and only use `add` to decide the `node.dtype`

zhuhaozhe committed:
Update on "enable data type propagation"


Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable data type propagation"


Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable data type propagation"


Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe committed:
Update on "enable data type propagation"


Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented:
@zhuhaozhe , please check the failed cases.

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "enable data type propagation"


Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable data type propagation"


Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge
